### PR TITLE
Refactor/ogmios 6 breaking changes

### PR DESCRIPTION
**Pull request by mirceahasegan**
# Context

Conway support requires [ogmios 6.0.0, which is a major rewrite](https://github.com/CardanoSolutions/ogmios/blob/master/architectural-decisions/accepted/017-api-version-6-major-rewrite.md#context).

# Proposed Solution
Update the @cardano-sdk/ogmios package such that the rewrite has a minor impact for the rest of the SDK.
Mappers/translators ogmiosToCore should use the updated Ogmios schema to produce the same Core types, so that the rewrite does not ripple throughout the SDK code and its users.

# TODO
- [x] Block
- [x] Transaction
- [x] Check if MIR certificate should be extracted from the new [Conway governance actions transaction](https://github.com/CardanoSolutions/ogmios/blob/master/architectural-decisions/accepted/017-api-version-6-major-rewrite.md#treasury-transfers) field and mapped to a core MIR certificate
- [x] Ensure [all changes](https://github.com/CardanoSolutions/ogmios/blob/master/architectural-decisions/accepted/017-api-version-6-major-rewrite.md#context) are accounted for in SDK

### CODE REVIEW

File: packages/ogmios/src/ogmiosToCore/tx.ts [hunk 1]
```diff
@@ -65,319 +53,305 @@ const mapPoolParameters = (poolParameters: Schema.PoolParameters): Cardano.PoolP
...
-  if ('stakeKeyRegistration' in certificate) {
-    return {
-      __typename: Cardano.CertificateType.StakeKeyRegistration,
-      stakeKeyHash: Crypto.Ed25519KeyHashHex(certificate.stakeKeyRegistration)
-    };
+  switch (certificate.type) {
+    case 'stakeDelegation':
+      return {
+        __typename: Cardano.CertificateType.StakeDelegation,
+        // REVIEW: how can the stakePool be undefined?
```
Comment (mirceahasegan) [inline packages/ogmios/src/ogmiosToCore/tx.ts:68]: REWIVEW: Clarify this scenario

File: packages/ogmios/src/ogmiosToCore/tx.ts [hunk 2]
```diff
@@ -65,319 +53,305 @@ const mapPoolParameters = (poolParameters: Schema.PoolParameters): Cardano.PoolP
...
-        ? Object.entries((tx as Schema.TxAlonzo).witness.redeemers).map(([key, value]) => mapRedeemer(key, value))
-        : undefined,
-      scripts: [...Object.values(tx.witness.scripts).map(mapScript)],
+      bootstrap: tx.signatories.map(mapBootstrapWitness),
+      ...(tx.datums && { datums: mapWitnessDatums({ datums: tx.datums }) }),
+      ...(tx.redeemers && { redeemers: Object.entries(tx.redeemers).map(([key, value]) => mapRedeemer(key, value)) }),
+      ...(tx.scripts && { scripts: [...Object.values(tx.scripts).map(mapScript)] }),
       signatures: new Map(
-        Object.entries(tx.witness.signatures).map(([key, value]) => [
+        // REVIEW: 'witness.bootstrap' and 'witness.signatures' were both flattened to 'signatories'. How should we handle this?
```
Comment (mirceahasegan) [inline packages/ogmios/src/ogmiosToCore/tx.ts:345]: REVIEW: clarify flattened 'witness.bootstrap' and 'witness.signatures'

File: packages/ogmios/src/ogmiosToCore/tx.ts [hunk 3]
```diff
@@ -65,319 +53,305 @@ const mapPoolParameters = (poolParameters: Schema.PoolParameters): Cardano.PoolP
...
-      // rewardAccount: certificate.moveInstantaneousRewards.rewards.
-      // Add test for it too.
-    } as Cardano.MirCertificate;
-  }
-  throw new NotImplementedError('Unknown certificate mapping');
+
+  // MIR certificates are now part of the proposals as gov actions
+  // https://github.com/CardanoSolutions/ogmios/blob/master/architectural-decisions/accepted/017-api-version-6-major-rewrite.md#treasury-transfers)
+
+  // TODO: translate code below from proposal into MIR certificate
```
Comment (mirceahasegan) [inline packages/ogmios/src/ogmiosToCore/tx.ts:110]: TODO: MIR certificate is now part of government actions. Check if we need to extract it for backwards compatibility

File: packages/ogmios/src/ogmiosToCore/tx.ts [hunk 4]
```diff
@@ -65,319 +53,305 @@ const mapPoolParameters = (poolParameters: Schema.PoolParameters): Cardano.PoolP
...
+    return obj;
+  } else if (Array.isArray(obj)) {
+    return obj.map(mapMetadatum);
+  } else if (typeof obj === 'object' && obj !== null) {
+    return new Map(Object.entries(obj).map(([k, v]) => [mapMetadatum(k), mapMetadatum(v)]));
   }
-  throw new NotImplementedError(
-    `Unknown metadatum type: ${typeof obj === 'object' ? Object.keys(obj).join(',') : obj}`
-  );
+  // TODO: deprecate Uint8Array union type from Cardano.Metadatum
```
Comment (mirceahasegan) [inline packages/ogmios/src/ogmiosToCore/tx.ts:233]: REVIEW: deprecate Uint8Array union type from Cardano.Metadatum

File: packages/ogmios/src/ogmiosToCore/tx.ts [hunk 5]
```diff
@@ -65,319 +53,305 @@ const mapPoolParameters = (poolParameters: Schema.PoolParameters): Cardano.PoolP
...
-    blob: data.body.blob
-      ? new Map(Object.entries(data.body.blob).map(([key, value]) => [BigInt(key), mapMetadatum(value)]))
-      : undefined,
-    scripts: data.body.scripts ? data.body.scripts.map(mapScript) : undefined
+    blob: data.labels
+      ? new Map(Object.entries(data.labels).map(([key, value]) => [BigInt(key), mapMetadatum(value)]))
+      : undefined
+    // Extra scripts have been moved to the scripts field and merged with witness scripts.
+    // https://github.com/CardanoSolutions/ogmios/blob/master/architectural-decisions/accepted/017-api-version-6-major-rewrite.md#transaction
+    // TODO: deprecate 'scripts' from auxiliaryData
```
Comment (mirceahasegan) [inline packages/ogmios/src/ogmiosToCore/tx.ts:252]: TODO: deprecate 'scripts' from auxiliaryData

File: packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts [hunk 6]
```diff
@@ -27,7 +27,7 @@ export interface InteractionContextProps {
    * 'OneTime' will close the connectino (and complete the interaction context observable)
    * after the 1st usage of emitted interaction context.
    */
-  interactionType: InteractionType;
+  interactionType: InteractionType; // REVIEW: Deprecate this? InteractionType is not used anymore in ogmios
```
Comment (rhyslbw) [inline packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts:30]: Yes, but probably do it as a separate commit as a cleanup step. So all interactions are "long running"? I'm guessing the availability of a HTTP API for "one-time" requests is the reason for the deprecation?

File: packages/ogmios/src/CardanoNode/queries.ts [hunk 7]
```diff
@@ -11,16 +12,19 @@ const wrapError = async <T>(query: () => Promise<T>) => {
...
     const eraSummaries = await client.eraSummaries();
     return eraSummaries.map((era) => eraSummary(era, systemStart));
   });
 
-export const queryGenesisParameters = (client: StateQueryClient, logger: Logger) =>
+export const queryGenesisParameters = (client: LedgerStateQuery.LedgerStateQueryClient, logger: Logger) =>
   wrapError(async () => {
     logger.info('Querying genesis parameters');
-    return genesis(await client.genesisConfig());
+    // REVIEW: The queryNetwork/genesis local-state-query now expects one era as argument (either 'byron', 'shelley' or 'alonzo')
```
Comment (rhyslbw) [inline packages/ogmios/src/CardanoNode/queries.ts:26]: This is a good improvement with Ogmios. Part of the follow up work we could expose this as a parameter.

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 1 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:68]: Could open an issue/discussion in ogmios repo to report/ask

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 3 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:110]: I'm guessing MIR is being deprecated?

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 4 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:233]: Yeah this is odd, worth asking. Also the map is now `{[string]: Metadatum}` which doesn't correspond to cddl model

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 5 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:252]: Agree with @AngelCastilloB here. Since there's information loss here we can't determine where it came from. Mapping to witness scripts is fine for our use case for now, but we should probably update our `OnChainTx` model later, because this mapping will be inaccurate representation.

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 2 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:345]: Bootstrap witness will have `chainCode` and `attributes`, so we can map this without loss

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 5 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:252]: Actually it says `The hash digests of those metadata-scripts are now listed as requiredExtraScripts`. Maybe it means that there's no information loss after all and we can check if script hash is present in that field to determine the correct mapping?

**Comment by rhyslbw** [comment]
This will need to be rebased on `feat/sanchonet-support` @mkazlauskas

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 4 above)
Comment (rhyslbw) [reply packages/ogmios/src/ogmiosToCore/tx.ts]: Using CBOR for metadata now

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 1 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:68]: @mirceahasegan

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 2 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:356]: @mirceahasegan apply the opposite filter to bootstrap witnesses

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 3 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:110]: We need to ask how MIR certificates are mapped during chain sync

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 5 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:252]: We can remove the `witness.scripts` and `auxiliaryData.scripts` in `OnChainTx` (keep in `Tx` type) type because we are not using it for anything. We can come back to this if/when we need to do anything with OnChainTx scripts

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 1 above)
Comment (mirceahasegan) [reply packages/ogmios/src/ogmiosToCore/tx.ts:68]: I opened [Q&A topic in ogmios](https://github.com/CardanoSolutions/ogmios/discussions/339)

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 3 above)
Comment (mirceahasegan) [reply packages/ogmios/src/ogmiosToCore/tx.ts:110]: asked here: https://github.com/CardanoSolutions/ogmios/discussions/340

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 5 above)
Comment (mirceahasegan) [reply packages/ogmios/src/ogmiosToCore/tx.ts]: Removed here: [3d41f58](https://github.com/input-output-hk/cardano-js-sdk/pull/927/commits/3d41f5853e0968fd4d33ebca4d4c2d5a5627d429)
@AngelCastilloB, please have a second look at this 😄 .

File: packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts (hunk 6 above)
Comment (mirceahasegan) [reply packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts:31]: Ogmios.createInteractionContext no longer expects an `interactionType`.

Comment (mirceahasegan) [reply packages/ogmios/src/CardanoNode/OgmiosObservableCardanoNode/createObservableInteractionContext.ts]: Done here: [6ab0ec8](https://github.com/input-output-hk/cardano-js-sdk/pull/927/commits/6ab0ec8f536b3d72f472eee75002d2771f62c30f)

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 3 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts:110]: So we probably need to map `GovernanceAction<TreasuryTransfer>` differently depending on era of the block. We're not doing anything with it right now so it's ok to create a separate ticket

**Review by mkazlauskas — APPROVED** [review]
Great work! :rocket:

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 3 above)
Comment (rhyslbw) [reply packages/ogmios/src/ogmiosToCore/tx.ts:146]: What's the action here?

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 5 above)
Comment (rhyslbw) [reply packages/ogmios/src/ogmiosToCore/tx.ts]: @mkazlauskas can you resolve this on behalf of @AngelCastilloB ?

**Review by rhyslbw — APPROVED** [review]
Fantastic work everyone involved, big milestone here 🚀

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 3 above)
Comment (mirceahasegan) [reply packages/ogmios/src/ogmiosToCore/tx.ts:146]: It is logged as a separate task

File: packages/ogmios/src/ogmiosToCore/tx.ts (hunk 5 above)
Comment (mkazlauskas) [reply packages/ogmios/src/ogmiosToCore/tx.ts]: Looks good to me. We may have to come back to this if we ever need to do something with OnChainTx scripts
